### ISSUE TITLE
Authenticator extensions {credProtect} are not a subset of requested extensions {}

### ISSUE THREAD

**Issue by mttcr** [Yubico/java-webauthn-server#92]
Hello,
I have the following error while trying to register a resident key:
```
com.yubico.webauthn.exception.RegistrationFailedException: java.lang.IllegalArgumentException: Authenticator extensions {credProtect} are not a subset of requested extensions {}.
	at com.yubico.webauthn.RelyingParty.finishRegistration(RelyingParty.java:446)
...
Caused by: java.lang.IllegalArgumentException: Authenticator extensions {credProtect} are not a subset of requested extensions {}.
	at com.yubico.webauthn.ExtensionsValidation.validate(ExtensionsValidation.java:61)
	at com.yubico.webauthn.FinishRegistrationSteps$Step12.validate(FinishRegistrationSteps.java:371)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.next(FinishRegistrationSteps.java:115)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps$Step.run(FinishRegistrationSteps.java:123)
	at com.yubico.webauthn.FinishRegistrationSteps.run(FinishRegistrationSteps.java:89)
	at com.yubico.webauthn.RelyingParty.finishRegistration(RelyingParty.java:444)
	... 40 more
```

I have the problem with Yubikey 5.2.6 and Windows 10 2004.
I don't have the problem with neither Yubikey 5.1.2 nor Windows 10 1909.

The creation options are
```json
{
	"attestation": "none",
	"authenticatorSelection": {
		"authenticatorAttachment": "cross-platform",
		"requireResidentKey": true,
		"userVerification": "required",
	},
	"challenge": [8, 185],
	"extensions": {},
	"pubKeyCredParams": [
		{"alg": -7, "type": "public-key"}
	],
	"rp": {
		"id": "localhost",
		"name": "xxx",
	},
	"user": {
		"displayName": "xxx",
		"id": [149, 41],
		"name": "xxx"
	}
}
```

The data sent back contains an extension:
![image](https://user-images.githubusercontent.com/8974534/96844423-c7ae6880-144f-11eb-8f9e-efb24e37c707.png)
![image](https://user-images.githubusercontent.com/8974534/96844464-d1d06700-144f-11eb-85d5-837a1ba66654.png)

Do you have any idea how to fix this behavior?

**Comment by emlun** [comment]
Ah, right. This is intended behaviour that has become a bug because of changes in the WebAuthn spec. In level 1, step 12 of [§7.1. Registering a New Credential](https://www.w3.org/TR/2019/PR-webauthn-20190117/#registering-a-new-credential) reads (emphasis added):

>12. Verify that the values of the client extension outputs in clientExtensionResults and the authenticator extension outputs in the extensions in authData are as expected, considering the client extension input values that were given as the extensions option in the create() call. In particular, **any extension identifier values** in the clientExtensionResults and the extensions in authData **MUST be also be present** as extension identifier values **in the extensions member of options**, i.e., **no extensions are present that were not requested**. In the general case, the meaning of "are as expected" is specific to the Relying Party and which extensions are in use.

In level 2 (which is still a work in progress), precisely because of the new `credProtect` extension, [the requirement to reject unsolicited extensions has been removed](https://www.w3.org/TR/webauthn-2/#sctn-registering-a-new-credential):

>17. Verify that the values of the client extension outputs in clientExtensionResults and the authenticator extension outputs in the extensions in authData are as expected, **considering** the client extension input values that were given in options.extensions and **any specific policy of the Relying Party regarding unsolicited extensions**, i.e., those that were not specified as part of options.extensions. In the general case, the meaning of "are as expected" is specific to the Relying Party and which extensions are in use.

This will require a rather simple update of the library to fix, but I'm afraid there's nothing you can do as a library user to work around it.

**Comment by emlun** [comment]
Oh, wait, I'd forgotten - there is the [`allowUnrequestedExtensions`](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/latest/com/yubico/webauthn/RelyingParty.RelyingPartyBuilder.html#allowUnrequestedExtensions(boolean)) setting on `RelyingParty`. It's disabled by default, but we should probably change it to be enabled by default. Anyway, try enabling it, that should fix your problem.

**Comment by mttcr** [comment]
It actually fixes the problem! Thanks a lot.

**Comment by emlun** [comment]
The default behaviour is now fixed in version 2.0.0. Thanks again!
